cp_diffdock_service.py

```diff
@@ -0,0 +1,188 @@
+import asyncio
+import json
+import logging
+import re
+import os
+from websockets.server import serve
+import concurrent.futures
+from argparse import ArgumentParser
+
+from cp_diffdock_protocol import DiffDockProtocol
+from cp_diffdock_api import DiffDockApi
+from cp_ws_helpers import extractWsAppMessage, formWsAppMessage, formCompletedMessage
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger('diffdock_service')
+
+#
+# This is a websocket server for servicing DiffDock requests.
+# This listens on a port and places incoming json requests into a queue, serviced by a number of workers.
+#
+
+async def handleRequest(websocket, queue):
+    """
+    Receive a request and put it into the queue.
+    Make sure we keep the websocket so we can send the response when docking finishes.
+    """
+    try:
+        message = await websocket.recv()
+    except Exception as ex:
+        log.error(f"Exception reading from websocket: {ex}. Can't handle this request.")
+        return
+
+    log.info("Received a request.")
+    # Optional special handling for messages from bfd-server or other Bioleap WsApps
+    # requestId is non-empty if it is from a wsApp, and will need to be included in responses.
+    requestId, cmdName, requestData = extractWsAppMessage(message)
+    requestContext = websocket, requestId
+
+    # Parse
+    try:
+        requestObj = DiffDockProtocol.Request.from_json(requestData)
+    except:
+        log.exception(f"Rejected a request:\nMessage: {message}\nrequestData: {requestData}")
+        await sendError(requestContext, f"Invalid request")
+        return
+
+    # Enqueue
+    log.info("handleRequest running a request...")
+    try:
+        # Put the request into the queue with a condition to resolve when it finishes.
+        # Condition protects the websocket from closing early.
+        condition = asyncio.Condition()
+        await queue.put((requestObj, requestContext, condition))
+        await sendStatus(requestContext, "Request received")
+        async with condition:
+            await condition.wait()
+    except Exception as ex:
+        log.exception(f"handleRequest encountered exception")
+        await sendError(requestContext, f"Service exception: {ex}")
+
+
+async def queueWorker(queue):
+    """
+    Handle docking requests from the queue
+    """
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        while True:
+            # Wait for an item from the queue
+            requestObj, requestContext, condition = await queue.get()
```

Comment:
Bizarrely, asyncio.Queue is explicitly documented as NOT being thread-safe. I think python's GIL (global interpreter lock) makes it safe (at least until the next python release or two changes that). But that feels iffy. I'm having trouble reasoning about 
- GIL
- python threads
- async tasks being resumed in any available thread
- multiple threads supposed to have released the GIL while running C++ (inference pytorch, numpy) 

Do the multiple diffdocks actually execute concurrently? I.e. with 5 worker threads do 5 diffdocks run 4-5x faster than just one?

If "it works" I have no complaints, just curiosity!

Reply:
The DiffDock workers appear to run concurrently, but slower because they share a single GPU resource.
I noticed occasionally one of the docks ends up with an empty results directly, which makes me wonder if somehow the DiffDock python code might be getting into trouble when multiple processes are accessing it. I am going to set the worker count to 1 for the first release of this service.